tagbot.py

```diff
@@ -0,0 +1,106 @@
+import gitlab
+import json
+import os
+import re
+import traceback
+
+re_repo = re.compile("Repository:.*/(.*/.*)")
+re_version = re.compile("Version:\\s*(v.*)")
+re_commit = re.compile("Commit:\\s*(.*)")
+
+merge = os.getenv("AUTOMATIC_MERGE", "").lower() == "true"
+registrator = int(os.environ["REGISTRATOR_ID"])
+token = os.environ["GITLAB_WEBHOOK_TOKEN"]
+client = gitlab.Gitlab(
+    os.environ["GITLAB_URL"], private_token=os.environ["GITLAB_API_TOKEN"]
+)
+
+
+def handler(evt, _ctx):
+    """Lambda entrypoint."""
+    try:
+        if get_in(evt, "headers", "X-Gitlab-Token") != token:
+            status, err = 403, "Invalid token"
+        else:
+            status, err = 200, handle_event(json.loads(evt.get("body", "{}")))
```

Comment:
Are there possibly exceptions we could `raise` in other functions that bubble up here to give a little more insight on where something failed? Could raise `RuntimeError`s in the other functions when something bad happens, and then the error messages would bubble up here.

Reply:
Or at least put the actual error message that gets raised in in the `err` error text?

Reply:
There are some exceptions that can be raised, yes. The way it currently works is basically:

- Any "validation" error goes to `err` as a simple message i.e. "this isn't a MR" and gets printed
- Any actual exception gets caught and printed via `traceback.print_exc()`

So any error that gets raised still does end up in the Cloudwatch logs, with a full stacktrace for exceptions.

Reply:
OH, yeah sorry I missed that traceback.print_exc()